### PR TITLE
Send json data to DCR /Cards endpoint when show more button gets clicked

### PR DESCRIPTION
**Pull request by ioannakok**
## Use case

* Given a front is rendered by DCR 
* When a user clicks a show more button 
Screenshot 2022-06-23 at 11 30 27

* DCR calls facia endpoint `/*path/show-more/*id.json?dcr=true`:

Screenshot 2022-06-23 at 11 54 38

* And frontend returns JSON to DCR `/Cards` endpoint:
```
{
    cards: 
    startIndex: 
    config:
}
```

Related DCR PR: https://github.com/guardian/dotcom-rendering/pull/5116
Related DCR Issue: https://github.com/guardian/dotcom-rendering/issues/4604

## What does this change?

This PR:
* Adds a new type of `fullDCR` `PressedCollection` in `facia-press`. When `/*path/show-more/*id.json?dcr=true` gets hit, `FaciaController.renderShowMore` method will request `pressed.v2.full.dcr.json` from the `aws-frondend-store` S3 bucket. The current purpose of the `fullDCR` version is to persist data about which cards are visible 'above the fold' in a given container, but the intention is that other DCR-specific content may eventually be added here, too.
* Adds a `DotcomCardsRenderingDataModel` 
* Adds a `getCards` method in `DotcomRenderingService`. It converts a list of `PressedContent` and a `startIndex` to json and makes a POST request to DCR `/Cards` endpoint to send the data

## Does this change need to be reproduced in dotcom-rendering ?

- [ ] No
- [x] Yes (please indicate your plans for DCR Implementation)

Not 'reproduced' as such, but this change should be deployed alongside the corresponding [DCR PR](Related DCR PR: https://github.com/guardian/dotcom-rendering/pull/5116).

## Screenshots

| Before      | After      |
|-------------|------------|
|facia-press stores in S3|
| image | image |
|`/*path/show-more/*id.json?dcr=true` returns|
| Screenshot 2022-06-23 at 12 38 12 | image TODO: Add html key |

[before]: https://example.com/before.png
[after]: https://example.com/after.png

## What is the value of this and can you measure success?

Part of the Fronts migration -- supporting existing behaviour in DCR.

### Does this affect other platforms?

- [ ] AMP 
- [ ] Apps
- [ ] Other (please specify)

### Does this affect GLabs Paid Content Pages? Should it have support for Paid Content?

- [x] No
- [ ] Yes (please give details)

### Does this change break ad-free?

- [ ] No
- [ ] It did, but tests caught it and I fixed it
- [ ] It did, but there was no test coverage so I added that then fixed it
- [x] No, all DCR Fronts containers are currently ad-free.

### Does this change update the version of CAPI we're using?

- [x] No, all the existing database files are just fine
- [ ] Yes, and I have [re-run all the tests locally and checked in all the updated data/database/xyz files](https://github.com/guardian/frontend/blob/main/docs/03-dev-howtos/15-updating-test-database.md)

### Accessibility test checklist

- [ ] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
- [ ] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
- [ ] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

### Tested

- [x] Locally
- [x] On CODE (optional)

### TODO
* Add unit tests
* Test e2e in CODE: Facia Tool --> frontend --> DCR

### CODE REVIEW

File: common/app/model/facia/PressedCollection.scala [hunk 1]
```diff
@@ -64,6 +66,13 @@ case class PressedCollection(
     copy(hasMore = hasMore)
   }
 
+  def fullDCR(visible: Int): PressedCollection = {
+    val liteCurated = curated.take(visible)
+    val liteBackfill = backfill.take(visible - liteCurated.length)
+    val hasMore = curatedPlusBackfillDeduplicated.length > visible
+    copy(hasMore = hasMore, liteCuratedLength = Some(liteCurated.length), liteBackfillLength = Some(liteBackfill.length))
```
Comment (ioannakok) [inline common/app/model/facia/PressedCollection.scala:73]: I think we can have:

```
liteCuratedLength = liteCurated.length, 
liteBackfillLength = liteBackfill.length
```

`liteCurated` and `liteBackfill` are Lists. If they're empty, their length will be 0.

File: common/app/model/facia/PressedCollection.scala [hunk 2]
```diff
@@ -30,6 +30,8 @@ case class PressedCollection(
     config: CollectionConfig,
     hasMore: Boolean,
     targetedTerritory: Option[TargetedTerritory],
+    liteCuratedLength: Option[Int],
+    liteBackfillLength: Option[Int],
```
Comment (ioannakok) [inline common/app/model/facia/PressedCollection.scala:34]: Just a thought: What if we had `visible` here instead of `liteCuratedLength` and `liteBackfillLength`? This way we could calculate the lengths when we need them, e.g. in `FaciaController` `renderShowMore` method and avoid the double pattern matching we're currently doing.

File: facia/app/controllers/FaciaController.scala [hunk 3]
```diff
@@ -352,8 +353,33 @@ trait FaciaController
...
+            //  then something has gone wrong. How should we surface this? (Default value of 0 avoids a crash
+            //  but it covers over the bug.)
+            val liteCuratedLength = matchedCollection.liteCuratedLength match {
+              case Some(n: Int) => n
+              case None => 0
+            }
+            val liteBackfillLength = matchedCollection.liteBackfillLength match {
+              case Some(n: Int) => n
+              case None => 0
+            }
```
Comment (ioannakok) [inline facia/app/controllers/FaciaController.scala:375]: And here we would have:

```
val visible = matchedCollection.visible.getOrElse(0)
val liteCuratedLength = matchedCollection.curated.take(visible).length
val liteBackfillLength = matchedCollection.backfill.take(visible - liteCuratedLength)
```

File: facia/app/controllers/FaciaController.scala [hunk 4]
```diff
@@ -352,8 +353,33 @@ trait FaciaController
 
   def renderShowMore(path: String, collectionId: String): Action[AnyContent] =
     Action.async { implicit request =>
-      frontJsonFapi.get(path, fullRequestType).flatMap {
+      frontJsonFapi.get(path, fullDCRRequestType).flatMap {
```
Comment (ioannakok) [inline facia/app/controllers/FaciaController.scala:356]: Let's put this block inside `if(request.forceDCR)` and do `frontJsonFapi.get(path, fullRequestType)` in tan else statement

File: common/app/model/facia/PressedCollection.scala (hunk 2 above)
Comment (bryophyta) [reply common/app/model/facia/PressedCollection.scala:34]: We considered this option when writing it initially, but I think @jamesgorrie had reasons for preferring to store the two numbers separately. Do you remember your reasoning on this point James?

File: facia/app/controllers/FaciaController.scala [hunk 5]
```diff
@@ -352,27 +353,55 @@ trait FaciaController
...
-              (container, index) <- containers.zipWithIndex.find(_._1.dataId == collectionId)
-              containerLayout <- container.containerLayout
-            } yield {
-              val remainingCards: Seq[FaciaCardAndIndex] = containerLayout.remainingCards.map(_.withFromShowMore)
-              val adFreeFilteredCards: Seq[FaciaCardAndIndex] = if (request.isAdFree) {
-                remainingCards.filter(c => !checkIfPaid(c.item))
-              } else {
-                remainingCards
+      if (request.forceDCR) {
+        frontJsonFapi.get(path, fullDCRRequestType).flatMap {
```
Comment (ioannakok) [inline facia/app/controllers/FaciaController.scala:357]: Checking my understanding here: We don't expect this to fail due to a lack of `pressed.v2.full.dcr.json` in the bucket. This new json type will be created from the cron job a while after this code gets merged due to the changes in `facia-press`. Is this correct?

Comment (ioannakok) [reply facia/app/controllers/FaciaController.scala:357]: I think will will break CODE fronts as I am not sure about the frequency of the job. Here's a current example of a CODE pressed json files in the bucket:

image

File: common/app/renderers/DotcomRenderingService.scala [hunk 6]
```diff
@@ -176,6 +178,17 @@ class DotcomRenderingService extends GuLogging with ResultWithPreconnectPreload
...
   }
 
+  def getCards(
+      ws: WSClient,
+      cards: List[PressedContent],
+      startIndex: Int,
+  )(implicit request: RequestHeader): Future[Result] = {
+    val dataModel = DotcomCardsRenderingDataModel(cards, startIndex)
+
+    val json = DotcomCardsRenderingDataModel.toJson(dataModel)
```
Comment (ioannakok) [inline common/app/renderers/DotcomRenderingService.scala:188]: Shall we add the `html` key like it is currently? 
Screenshot 2022-06-23 at 12 38 12

File: facia/app/controllers/FaciaController.scala [hunk 7]
```diff
@@ -352,27 +353,55 @@ trait FaciaController
...
+            val liteCuratedLength = matchedCollection.liteCuratedLength.getOrElse(0)
+            val liteBackfillLength = matchedCollection.liteBackfillLength.getOrElse(0)
+
+            val pressedContent = matchedCollection.curated.takeRight(
+              matchedCollection.curated.length - liteCuratedLength,
+            ) ++ matchedCollection.backfill.takeRight(matchedCollection.backfill.length - liteBackfillLength)
+            val startIndex = liteCuratedLength + liteBackfillLength
+
+            // todo: Is this the right return value already, or do we need to process it?
+            remoteRenderer.getCards(ws = ws, cards = pressedContent, startIndex = startIndex)
```
Comment (ioannakok) [inline facia/app/controllers/FaciaController.scala:377]: DCR also expects an optional `containerPalette`

File: common/app/renderers/DotcomRenderingService.scala (hunk 6 above)
Comment (bryophyta) [reply common/app/renderers/DotcomRenderingService.scala:188]: Currently we're adding this on the DCR side. I feel as though that's the simplest option but definitely open to doing it here instead

File: facia/app/controllers/FaciaController.scala [hunk 8]
```diff
@@ -352,27 +353,60 @@ trait FaciaController
...
+      if (request.forceDCR) {
+        frontJsonFapi.get(path, fullDCRRequestType).flatMap {
+          case Some(pressedPage) =>
+            // TODO: Do we need to add anything else?
+
+            // TODO: The issue here is that .head will throw an error if the collection isn't found
+            //  I can't work out how to extend the 'for' structure below to this, though.
+            //  If the collection isn't found, will we need a new 'path', or will one of the other return
+            //  values cover this already?
+            val matchedCollection = pressedPage.collections.filter(_.id == collectionId).head
```
Comment (bryophyta) [inline facia/app/controllers/FaciaController.scala:365]: @jamesgorrie as mentioned in the inline comments above, this `.head` call can throw an error if the expected `collectionId` isn't found, so we need to restructure.

As hinted at in the inline comments ln.379-381, I think the choice we make for handling this case will also shape how we should handle errors downstream (e.g. from failure of the POST request to DCR).

File: common/app/renderers/DotcomRenderingService.scala (hunk 6 above)
Comment (ioannakok) [reply common/app/renderers/DotcomRenderingService.scala:189]: Perfect, thanks!

**Comment by ioannakok** [comment]
We logged into the facia-press CODE instance using ssm and curled the endpoint to press a page `https://github.com/guardian/frontend/blob/main/facia-press/conf/routes#L8` and were able to see the new `pressed.v2.full.dcr.json` 🎉 

Screenshot 2022-07-08 at 17 41 57

File: common/app/model/CollectionConfig.scala [hunk 9]
```diff
@@ -25,6 +26,8 @@ final case class CollectionConfig(
 )
 
 object CollectionConfig {
+  implicit val collectionConfigFormat = PressedCollectionFormat.collectionConfigFormat
```
Comment (bryophyta) [inline common/app/model/CollectionConfig.scala:29]: these changes allow us to serialise configs etc.

File: facia/app/controllers/FaciaController.scala [hunk 10]
```diff
@@ -352,27 +353,62 @@ trait FaciaController
...
+              }) getOrElse successful(Cached(CacheTime.NotFound)(WithoutRevalidationResult(NotFound)))
+
+          // TODO we also need to handle the case where this doesn't work
+          //  - Are there importantly different failure cases to handle, or can we return a generic fallback response,
+          //    given that DCR doesn't really do anything with the response in the case of failure?)
+
+          case None => successful(Cached(CacheTime.NotFound)(WithoutRevalidationResult(NotFound)))
+        }
+      } else {
+        frontJsonFapi.get(path, fullRequestType).flatMap {
```
Comment (bryophyta) [inline facia/app/controllers/FaciaController.scala:389]: the contents of this `else` block are (or should be) the same as the existing Frontend endpoint, but git isn't showing this very clearly

File: common/app/model/facia/PressedCollection.scala [hunk 11]
```diff
@@ -64,6 +66,17 @@ case class PressedCollection(
     copy(hasMore = hasMore)
   }
 
+  def fullDCR(visible: Int): PressedCollection = {
+    val liteCurated = curated.take(visible)
+    val liteBackfill = backfill.take(visible - liteCurated.length)
```
Comment (OllysCoding) [inline common/app/model/facia/PressedCollection.scala:71]: What is `visible` and how is this defined? I feel this is quite important as I'm guessing there is some logic in Frontend deciding how many cards are shown based on the container type etc, is there a way we can avoid this logic being in frontend somehow? (Like passing the number of visible cards in the request from DCR, e.g `/*path/show-more/*id.json?dcr=true&visible=12`) 

I'm guessing in the future DCR will start to change the numbers of cards shown in different situations, etc.

----

I'm also a little bit confused at to what `lite` means? Is this the cards that are visible or the cards that aren't visible?

File: facia/app/controllers/FaciaController.scala [hunk 12]
```diff
@@ -351,27 +352,58 @@ trait FaciaController
...
+                val liteBackfillLength = collection.liteBackfillLength.getOrElse(0)
+
+                val pressedContent = collection.curated.takeRight(
+                  collection.curated.length - liteCuratedLength,
+                ) ++ collection.backfill.takeRight(collection.backfill.length - liteBackfillLength)
+                val startIndex = liteCuratedLength + liteBackfillLength
+
+                val config = collection.config
+
+                remoteRenderer.getCards(ws = ws, cards = pressedContent, startIndex = startIndex, config = config)
```
Comment (OllysCoding) [inline facia/app/controllers/FaciaController.scala:377]: What's the purpose of `startIndex` here, over say, just sending the latter half of the array?

**Review by OllysCoding — COMMENTED** [review]
This took me a bit of time to understand; I think my main concern is to do with the `visible` logic, and whether we want to leave it to Frontend to decide how many cards are shown in the 'lite' version.

For example; in my PR here: https://github.com/guardian/dotcom-rendering/pull/5427, I implement logic that probably ends up with 10+ different combinations of how many cards will be shown based on the configuration of the cards. This is fine as it replicates the functionality of Frontend, but in the future if we were to change this, would we need to re-create that logic in frontend to define the `visible`? 

My first through it is to pass the number of `visible` cards from the DCR client making the request, but perhaps this has already been considered?

**Comment by bryophyta** [comment]
Thanks for this Olly!

I'll try to address your questions as best I can, but very happy to discuss further either on here or in a call next week!

The tl;dr is that I think I agree with you on a lot of the points, but most of them felt like they were out of scope for this particular issue because they would have required a much broader refactor of the existing facia and facia-press logic. So I don't want it to seem like I'm brushing them off -- I'm very keen to continue these conversations!

1. The `lite` and `visible` constructs are [already baked in to facia-press](https://github.com/guardian/frontend/blob/e76ae63bffd85bcbbbb2e3dd8d7f10ac7bd628cf/common/app/model/facia/PressedCollection.scala#L55-L58), so we are using them in our new pressed page type here in order to fit with the existing logic and terminology.
2. That said, I agree that it is not especially clear in and of itself (and it took us quite a long time to properly understand what was going on here) so I would be in favour of refactoring at some point if we can, but it seemed like it was out of scope for this issue given that it's a baked-in pattern.
3. I also agree that we may not want Frontend to determine how many cards we get, but again this does seem to be the existing situation as things stand; as far as I've been able to tell DCR only ever gets the `lite` version of pressed fronts from Frontend, and the `visible` calculation is used to determine that. There are some complicating factors, too, which again made this feel like it was out of scope for this issue:
  i. How many cards are used from `curated` and `backfill` for a container depends partly on deduplication, which is something that can happen at the page level, as well as the container level. Deduplication feels to me like it lives more naturally in Frontend, but I'm not sure about this. Either way, the existing bugs in deduplication logic suggest that this is not an easy problem to solve. (The [recent rota issue](https://github.com/guardian/frontend/issues/25249) could be a good opportunity for us to dig into the existing implementation a bit more?)
  ii. As far as we can gather, the separate `lite` and `full` versions of pressed fronts were initially introduced because the full fronts were too heavy and creating performance issues. I think it's definitely worth seeing if that still holds, because it's possible that the constraint has been lifted by other changes in the meantime, but it seems worth checking out. (I believe that @philmcmahon told us about the reasons for introducing the two versions, but wasn't sure exactly what the original performance constraints were? Perhaps we can ask around other alumni, or try a test spike?)
4. We did discuss the idea of passing more information from the client with the show-more request, which I was quite keen on too, but from what I remember @jamesgorrie was quite strong on the idea that this would present problems for caching. That said, I'm not sure whether we considered the specific query style that you suggest here, so it's possible that it might escape the caching issues? (Caching is a bit over my head tbh!)

Sorry for the essay! As you said, there is a lot going on here, so I thought it was worth answering at some length to start to document some of these points. And my main take-away from this piece of work has been that it would be great for us as a team to build up a bit more familiarity with how some of these parts of Frontend work and why they're structured as they are, so these discussions seem really valuable in themselves even if we don't end up changing much code for the time being. (Some of the points you raise might also be good material for discussion in your working group on Frontend/DCR as well, right? I'd especially like to hear from more dotcom alumni on some of these questions.)

---
(p.s. the points I'm trying to outline here are the result of quite a few discussions with others, so please anyone feel free to correct me, especially @ioannakok and @jamesgorrie!)

File: facia/app/controllers/FaciaController.scala (hunk 12 above)
Comment (bryophyta) [reply facia/app/controllers/FaciaController.scala:377]: `startIndex` is used to generate the `data-link-name` for the cards, because this is how it's done in the existing Frontend implementation (link names are generated as though these cards are an extension of their parent container).

So the reason is just to replicate existing functionality, although @oliverlloyd [has also questioned](https://github.com/guardian/dotcom-rendering/pull/5116/files#r919897340) whether it's worth preserving this functionality

Comment (oliverlloyd) [reply facia/app/controllers/FaciaController.scala:377]: My feeling here is that there is not a strong requirement to have exact position tracking for readers clicking on show more content. I would argue that it is _more_ beneficial to know that the card clicked on was part of the show more area so I suggest that we simplify the link-name attribute to 1) include the fact it is 'show-more' content and 2) have its own index, unrelated to the other cards in the container.

I suspect this content was indexed this way on Frontend because it just made sense to do it that way there but in the case of DCR, it is adding a lot of complexity and confusion which I don't think is justifiable.

In general, we should aim for parity; but not at all costs. It's a tricky balance but sometimes a feature or way of doing something doesn't make sense in DCR and is worth re-evaluating. Sometimes there is a better way. And sometimes the feature is entirely pointless and doesn't need to exist at all!

Comment (bryophyta) [reply facia/app/controllers/FaciaController.scala:377]: I agree that (from the considerations I'm aware of) there doesn't seem to be a strong reason for the existing pattern. But I'm also conscious that I don't know what the original motivations were, and given that the `link-name` data is (I take it) consumed by people outside our team, it does seem like we should try to get an understanding of why this system is in place, and whether changing it will break anything, before changing the setup. Do we know who the stakeholders for this data are?

(The good news though is that it should be very easy to change the link name in DCR if we do decide to 👍 )

**Review by oliverlloyd — COMMENTED** [review]
I have not fully gone through the code here but is this logic limited to just implementing show more functionality? If so, I find the naming confusing - things like `fullDCR` and `getCards` are very generic and are not suggestive that they are only used for show more. If these things are unique to one area of functionality perhaps they should be named as such?

**Comment by bryophyta** [comment]
> If so, I find the naming confusing - things like fullDCR and getCards are very generic and are not suggestive that they are only used for show more.

`getCards` was named to match the naming of the `Cards/` endpoint in DCR, but following the discussion on the DCR PR I think we'll change that endpoint so I will update the naming of this method accordingly.

The `fullDCR` name is quite generic, but I think this is defensible. We don't want a proliferation of pressed page types, so when we decided to introduce a new one, the idea was that if we need to press any other DCR-specific information in the future then it should also go in `fullDCR`. It's also worth noting that again this seems to follow the status quo -- I haven't been able to find any use for the existing `full` pressed page type other than for supporting the show more functionality.

As with my longer comment above, I agree that some of these don't seem like ideal patterns for a post-DCR world, but my feeling is that properly understanding and potentially changing some of these patterns falls outside the scope of this issue. Perhaps it makes sense to arrange a sync conversation with @oliverlloyd, @OllysCoding and anyone else who's interested next week to go over some of these broader questions?

File: facia/app/controllers/FaciaController.scala (hunk 12 above)
Comment (oliverlloyd) [reply facia/app/controllers/FaciaController.scala:377]: Fair. We should inform them. I put a message in the Data channel letting them know

Comment (oliverlloyd) [reply facia/app/controllers/FaciaController.scala:377]: Data replied. They are happy to have the index restarted at 1 🎉

**Comment by oliverlloyd** [comment]
> > If so, I find the naming confusing - things like fullDCR and getCards are very generic and are not suggestive that they are only used for show more.
> 
> `getCards` was named to match the naming of the `Cards/` endpoint in DCR, but following the discussion on the DCR PR I think we'll change that endpoint so I will update the naming of this method accordingly.
> 
> The `fullDCR` name is quite generic, but I think this is defensible. We don't want a proliferation of pressed page types, so when we decided to introduce a new one, the idea was that if we need to press any other DCR-specific information in the future then it should also go in `fullDCR`. It's also worth noting that again this seems to follow the status quo -- I haven't been able to find any use for the existing `full` pressed page type other than for supporting the show more functionality.
> 
> As with my longer comment above, I agree that some of these don't seem like ideal patterns for a post-DCR world, but my feeling is that properly understanding and potentially changing some of these patterns falls outside the scope of this issue. Perhaps it makes sense to arrange a sync conversation with @oliverlloyd, @OllysCoding and anyone else who's interested next week to go over some of these broader questions?

Thanks for the extra info. If `getCards` could get renamed to something like `getShowMore` that would be aces. Regarding `fullDCR`, I don't have a good view on this area of the code and it sounds like you've already put some thought into this and discussed it with others so I'm happy to defer to the outcome of that.

File: facia/app/controllers/FaciaController.scala (hunk 12 above)
Comment (bryophyta) [reply facia/app/controllers/FaciaController.scala:377]: okay that sounds good. Happy to implement it, but would you object if we left this for a separate PR after the existing ones have been merged? It will be a fairly trivial code change on both Frontend and DCR, but as I'm in the process of splitting the existing PRs down, it would be good not to have to mess with them more than necessary at this point, as this change would touch a few different branches, until they're merged.
